lib/KettleRequest.js

```diff
@@ -0,0 +1,282 @@
+kettle.wrapCallback = function (callback) {
+    var request = kettle.getCurrentRequest();
+    return kettle.withRequest(request, callback);
+};
+
+fluid.defaults("kettle.request", {
+    gradeNames: ["fluid.component"],
+    mergePolicy: {
+        requestMiddleware: "noexpand"
+    },
+    invokers: {
+        handleRequest: "fluid.notImplemented",
+        handleFullRequest: "fluid.notImplemented"
+    },
+    events: {
+        onHandle: null,  // Fired in order to execute its listener, handleRequest, which invokes the user's handler
+        onSuccess: null, // A convenient proxy for the main request handler to report disposition (equivalent to handlerPromise)
+        onError: null,  // A convenient proxy for the main request handler to report disposition (equivalent to handlerPromise)
+        onRequestEnd: null,
+        onRequestSuccess: null, // Overall disposition of the entire request - handler actually sends response
+        onRequestError: null // Overall disposition of the entire request - handler actually sends response
+    },
+    // sourced from dynamic args
+    req: "{arguments}.1",
+    res: "{arguments}.2",
+    next: "{arguments}.3",
+    members: {
+        req: "{that}.options.req", // TODO: can't supply these direct from args because of framework members merging bug - get array instead
+        res: "{that}.options.res",
+        next: "{that}.options.next",
+        /* Construct a "promisified" proxy for the handling of this request. It will
+         * have handlers registered which forward to this request's success and error events.
+         * The promise may be rejected or resolved by the user to represent that disposition
+         * for the overall request */
+        handlerPromise: "@expand:fluid.promise()"
+    },
+    listeners: {
+        "onCreate.activate": "kettle.request.activate",
+        "onHandle.handleRequest": {
+            funcName: "kettle.request.handleRequest",
+            args: "{that}"
+        },
+        "onSuccess.forward": "kettle.request.forwardPromise(resolve, {that}.handlerPromise, {arguments}.0)",
+        "onError.forward": "kettle.request.forwardPromise(reject, {that}.handlerPromise, {arguments}.0)"
+    }
+});
+
+// A handler which reports a 404 if the request has not already been handled (e.g. by some middleware, e.g. static)
+kettle.request.notFoundHandler = function (request) {
+    if (!request.handlerPromise.disposition) {
+        request.handlerPromise.reject({statusCode: 404, message: "Cannot " + request.req.method + " " + request.req.originalUrl});
+    }
+};
+
+kettle.request.forwardPromise = function (method, promise, val) {
+    if (!promise.disposition) {
+        promise[method](val);
+    } else {
+        // Don't use fluid.fail here to avoid infinite triggering of errors
+        fluid.log("Error in forwarding result ", val, " to promise " + method + ": promise has already received " + promise.disposition);
```

Comment:
Istanbul claims this line is never hit in the tests, which I really hope is not true.